feature/s3/transfermanager/api_op_PutObject.go

```diff
@@ -0,0 +1,1077 @@
+package transfermanager
+
+import (
+	"bytes"
+	"context"
+	"fmt"
+	"io"
+	"net/http"
+	"sort"
+	"sync"
+	"time"
+
+	"github.com/aws/aws-sdk-go-v2/aws"
+	"github.com/aws/aws-sdk-go-v2/aws/middleware"
+	"github.com/aws/aws-sdk-go-v2/service/s3"
+	"github.com/aws/aws-sdk-go-v2/service/s3/types"
+	smithymiddleware "github.com/aws/smithy-go/middleware"
+)
+
+// A MultipartUploadError wraps a failed S3 multipart upload. An error returned
+// will satisfy this interface when a multi part upload failed to upload all
+// chucks to S3. In the case of a failure the UploadID is needed to operate on
+// the chunks, if any, which were uploaded.
+//
+// Example:
+//
+//	u := manager.NewUploader(client)
+//	output, err := u.upload(context.Background(), input)
+//	if err != nil {
+//		var multierr manager.MultiUploadFailure
+//		if errors.As(err, &multierr) {
+//			fmt.Printf("upload failure UploadID=%s, %s\n", multierr.UploadID(), multierr.Error())
+//		} else {
+//			fmt.Printf("upload failure, %s\n", err.Error())
+//		}
+//	}
+type MultipartUploadError interface {
+	error
+
+	// UploadID returns the upload id for the S3 multipart upload that failed.
+	UploadID() string
+}
+
+// A multiUploadError wraps the upload ID of a failed s3 multipart upload.
+// Composed of BaseError for code, message, and original error
+//
+// Should be used for an error that occurred failing a S3 multipart upload,
+// and a upload ID is available. If an uploadID is not available a more relevant
+type multipartUploadError struct {
+	err error
+
+	// ID for multipart upload which failed.
+	uploadID string
+}
+
+// batchItemError returns the string representation of the error.
+//
+// # See apierr.BaseError ErrorWithExtra for output format
+//
+// Satisfies the error interface.
+func (m *multipartUploadError) Error() string {
+	var extra string
+	if m.err != nil {
+		extra = fmt.Sprintf(", cause: %s", m.err.Error())
+	}
+	return fmt.Sprintf("upload multipart failed, upload id: %s%s", m.uploadID, extra)
+}
+
+// Unwrap returns the underlying error that cause the upload failure
+func (m *multipartUploadError) Unwrap() error {
+	return m.err
+}
+
+// UploadID returns the id of the S3 upload which failed.
+func (m *multipartUploadError) UploadID() string {
+	return m.uploadID
+}
+
+// PutObjectInput represents a request to the Upload() call.
+type PutObjectInput struct {
```

Comment:
Expand the comment and mention that this is an object that is the common field of `PutObject` and `CreateMultiPartUpload`.

I'm not sure about calling this `PutObjectInput`. On one hand is likely what you'd look for when using the transfer manager. However, it is confusing that it's almost exactly the same as `PutObjectInput` on S3 but not really.

Reply:
If we're trying to match what our normal clients look like, then it _should_ be PutObjectInput, since the operation is PutObject.

I'm fine with that naming since this is basically "put an object", where multipart split is an optimization therein.

Reply:
> Expand the comment and mention that this is an object that is the common field of PutObject and CreateMultiPartUpload.

This remains unaddressed.